### DISCUSSION TITLE
How can I get a field offset from a Class for managed ref?

### DISCUSSION THREAD

**Discussion by cyraid** [dotnet/runtime#83919] [Q&A]
Hi there,

Do field offsets get moved around? I noticed there was a NotImplemented.ByDesign if I recall poking around the source. Since there's no guarantee the internal metadata for a Class will stay the same, how would I get the field offset myself? (by just having access to typeof(ClassName))

Edit: Also note that the class is unstructured, so no [FieldOffset] and such (so I can't use Marshal.OffsetOf)

**Comment by cihatsolak** [comment]
Hi, field offsets can change, so you shouldn't rely on them being fixed. If you absolutely need to access the field of an unmanaged reference and the class is unconfigured (i.e. no [FieldOffset] attribute), you can use the System.Runtime.CompilerServices.Unsafe class to get the field offset at runtime. The Unsafe.OffsetOf method returns an IntPtr that represents the byte offset of the specified field in the type, which can be converted to a long value using the ToInt64 method.

**Reply by cyraid** [reply]
Hi @cihatsolak , thanks for the reply. ^_^ So if you have 3 object instances, all 3 objects can have different field offsets?

**Reply by cyraid** [reply]
Also, I don't see access to Unsafe.OffsetOf, is there an extension? (I'm using C# 7.0, .net core) .. Also, judging by these other conversations I'm reading: https://github.com/dotnet/runtime/issues/28001
https://github.com/dotnet/runtime/issues/43285 The offsets don't seem to change? And how would the field offset change from instance to instance? The C# runtime would have to have a table of offsets internally for each object instance (which would waste a lot of space). I don't think this seems to be the case, and think there's a field offset order determined at ~~compile time~~ after compile / while running (as in the field offsets aren't different per instance)

**Reply by Clockwork-Muse** [reply]
> So if you have 3 object instances, all 3 objects can have different field offsets?

No, it's going to vary at compile time (and maybe JIT?  Dunno there).

**Reply by cyraid** [reply]
Oh sorry I meant, when it's running, it has the field offsets finalized. But there's discussions on to let a possible get field offset like: RuntimeHelper.GetFieldOffset(RuntimeFieldHandle) hopefully (or some other method). https://github.com/dotnet/runtime/issues/45152

**Reply by cyraid** [reply]
But I know now it's realistically impossible (without using .emit, or undocumented abi access like internal table, etc) to get field offsets with managed references, but if the object/struct doesn't have any non-blittable types you can still use pinning and get the field offsets that way (at runtime).

**Reply by Clockwork-Muse** [reply]
Perhaps the better question is, **why** do you want this information?  What are you planning on doing with it?

**Reply by cihatsolak** [reply]
> Perhaps the better question is, **why** do you want this information? What are you planning on doing with it?

thank you for this good question

**Reply by cyraid** [reply]
Yup, good question.

Currently developing a virtual runtime that executes instructions (there are requirements to why I can't actually emit IL or bytecode, or use reflection), and it has to be NativeAOT compatible. The runtime executes functions from a code pointer which acts like the instruction pointer for a CPU. Having a delegate is too slow, which is why it's nice to have C# function pointers (which is put in a different question, as I need some functionality from there too).

The runtime accesses fields from instances, which need to be really fast. Currently I bind the fields when the Object is instantiated through a nice unsafe __makeref() pointer->pointer trick, then stores the offsets and doesn't need to do that after another Object of the same type is instantiated. Though, it would be nice if I could do this upfront.

Edit: Funny story, I was actually making the VM/runtime in C/C++, and it turns out I actually optimized the C# so much that it actually runs FASTER than C/C++! (in NativeAOT mode). Even when the instructions are similar. So good job to the C# folk. So now can just focus on all the work in C#. :D

**Reply by Clockwork-Muse** [reply]
.... this _sounds like_ you're trying to dodge around a "no dynamic code" restriction (although NativeAOT limits some of that anyways).

If you're using NativeAOT, though, I'm not sure what's going to happen things like those field offsets.,,

**Reply by cyraid** [reply]
> .... this _sounds like_ you're trying to dodge around a "no dynamic code" restriction (although NativeAOT limits some of that anyways).
> 
> If you're using NativeAOT, though, I'm not sure what's going to happen things like those field offsets.,,

Works for me quite nicely in NativeAOT mode. :) Just would like to do it not so 'hacky'. Now if only there was a way to convert delegates to managed function pointers. As for the fields go, I have it setup so a Source Generator I develop in the future can make a lot of the code I'm boiler plating currently.
